src/autotester/server/utils/tests/file_management_test.py

```diff
@@ -0,0 +1,333 @@
+        assert len(list_fd_bef_move) == len(list_fd_after_move)
+
+    def test_dir_has_onefile(self, empty_dir):
+        """
+        When the Source Directory has only one file
+        """
+        source_dir = tempfile.mkdtemp()
+        fd, source_file = tempfile.mkstemp(dir=source_dir)
+        dest_dir = empty_dir
+        move_tree(source_dir, dest_dir)
+        moved_file = os.path.join(dest_dir, os.path.basename(source_file))
+        assert os.path.exists(moved_file) and not os.path.exists(source_file)
+
+    def test_dir_has_onedir(self, empty_dir, dir_has_onedir):
+        """
+        When the Dirctory has only one sub directory
+        """
+        source_dir = tempfile.mkdtemp()
+        sub_dir = tempfile.mkdtemp(dir=source_dir)
+        dest_dir = empty_dir
+        move_tree(source_dir, dest_dir)
+        moved_dir = os.path.join(dest_dir, os.path.basename(sub_dir))
+        assert os.path.exists(moved_dir) and not os.path.exists(sub_dir)
+
+    def test_dir_has_nestedfd(self, empty_dir, nested_fd):
+        """
+        When the files are nested in subdirectories more than 2 directories deep
+        """
+        dest_dir = empty_dir
+        source_dir = tempfile.mkdtemp()
+        sub_dir1 = tempfile.mkdtemp(dir=source_dir)
+        sub_dir2 = tempfile.mkdtemp(dir=sub_dir1)
+        fd, source_file = tempfile.mkstemp(dir=sub_dir2)
+        move_tree(source_dir, dest_dir)
+        moved_dir1 = os.path.join(dest_dir, os.path.basename(sub_dir1))
+        moved_dir2 = os.path.join(moved_dir1, os.path.basename(sub_dir2))
+        moved_file = os.path.join(moved_dir2, os.path.basename(source_file))
+        assert os.path.exists(moved_dir1) and not os.path.exists(sub_dir1)
+        assert os.path.exists(moved_dir2)
+        assert os.path.exists(moved_file)
+
+
+class TestFdOpen:
+    def test_open_dir(self):
+        """
+        Checks whether two file descriptors are pointing to the same directory
+        """
+        with tempfile.TemporaryDirectory() as dir:
+            dir_fd = os.open(dir, os.O_RDONLY)
+            with fd_open(dir) as fdd:
+                same_dir = os.path.sameopenfile(fdd, dir_fd)
+            assert same_dir
+
+    def test_open_file(self):
+        """
+        Checks whether two file descriptors are pointing to the same file
+        """
+        with tempfile.NamedTemporaryFile() as file:
+            file_fd = os.open(file.name, os.O_RDONLY)
+            with fd_open(file.name) as fdf:
+                same_file = os.path.sameopenfile(fdf, file_fd)
+            assert same_file
+
+    def test_close(self):
+        """
+        Checks whether the file or directory is closed
+        """
+        with tempfile.TemporaryDirectory() as dir:
+            with fd_open(dir) as fdd:
+                dir_fd = fdd
+            with pytest.raises(Exception):
+                os.close(dir_fd)
+
+
+def test_copy_test_script_files(dir_has_onefile, redis, tmp_script_dir):
```

Comment:
If you're going to use fixtures for your tests, you should use the fixtures in the test. 

- `dir_has_one_file` is required to setup your tests
- `redis` is not required because it has `autouse=True` specified and you don't refer to the `redis` variable in your test
- `tmp_script_dir` is required but you are not using it in your test even though you should be. 

Please review the code below and see how the changes I'm proposing mean your test still runs properly but the test itself is simpler and all fixtures are used:

```python
def test_copy_test_script_files(dir_has_onefile, tmp_script_dir):
    """
    Checks whether the test script files are copied to the destination directory
    """
    markus_address = "http://localhost:3000/csc108"
    assignment_id = 1
    tests_path, test_file = dir_has_onefile
    copy_test_script_files(markus_address, assignment_id, tests_path)
    test_script_dir = os.path.join(tmp_script_dir, FILES_DIRNAME)
    copied_file = os.path.join(test_script_dir, test_file)
    assert os.path.exists(copied_file)
```

I would like you to also make this test into a class (like you've done for all the other tests above) and include additional tests which check if files/directories are copied properly if there are multiple files/directories (like you've done for all the other tests above).